File: go/nemo_relay/adaptive.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package nemo_relay

import "encoding/json"

// AdaptivePluginKind is the top-level plugin kind used by the adaptive component.
const AdaptivePluginKind = "adaptive"

// AdaptiveConfig is the canonical Go shape for the adaptive plugin config document.
type AdaptiveConfig struct {
	Version         uint32                 `json:"version,omitempty"`
	AgentID         string                 `json:"agent_id,omitempty"`
	State           *AdaptiveStateConfig   `json:"state,omitempty"`
	Telemetry       *TelemetryConfig       `json:"telemetry,omitempty"`
	AdaptiveHints   *AdaptiveHintsConfig   `json:"adaptive_hints,omitempty"`
	ToolParallelism *ToolParallelismConfig `json:"tool_parallelism,omitempty"`
	Acg             *AcgConfig             `json:"acg,omitempty"`
	ResponseCache   *ResponseCacheConfig   `json:"response_cache,omitempty"`
	Policy          *ConfigPolicy          `json:"policy,omitempty"`
}

// AdaptiveStateConfig selects the adaptive state backend.
type AdaptiveStateConfig struct {
	Backend AdaptiveBackendSpec `json:"backend"`
}

// AdaptiveBackendSpec selects the backend kind and backend-specific config.
type AdaptiveBackendSpec struct {
	Kind   string         `json:"kind"`
	Config map[string]any `json:"config,omitempty"`
}

// TelemetryConfig configures the built-in adaptive telemetry subscriber and learners.
type TelemetryConfig struct {
	SubscriberName string   `json:"subscriber_name,omitempty"`
	Learners       []string `json:"learners,omitempty"`
}

// AdaptiveHintsConfig configures built-in LLM request hint injection.
type AdaptiveHintsConfig struct {
	Priority       int32  `json:"priority,omitempty"`
	BreakChain     bool   `json:"break_chain,omitempty"`
	InjectHeader   bool   `json:"inject_header,omitempty"`
	InjectBodyPath string `json:"inject_body_path,omitempty"`
}

// ToolParallelismConfig configures built-in adaptive tool scheduling.
type ToolParallelismConfig struct {
	Priority int32  `json:"priority,omitempty"`
	Mode     string `json:"mode,omitempty"`
}

// AcgStabilityThresholds configures prompt stability classification thresholds.
type AcgStabilityThresholds struct {
	StableThreshold                  float64 `json:"stable_threshold,omitempty"`
	SemiStableThreshold              float64 `json:"semi_stable_threshold,omitempty"`
	MinObservationsForFullConfidence uint32  `json:"min_observations_for_full_confidence,omitempty"`
}

// AcgConfig configures the adaptive cache governor.
type AcgConfig struct {
	Provider            string                  `json:"provider,omitempty"`
	ObservationWindow   uint32                  `json:"observation_window,omitempty"`
	Priority            int32                   `json:"priority,omitempty"`
	StabilityThresholds *AcgStabilityThresholds `json:"stability_thresholds,omitempty"`
}

// ResponseCacheConfig configures the opt-in LLM response cache: a section
// of the adaptive config (a sibling to acg/adaptive_hints/tool_parallelism), not a
// standalone plugin kind. The Rust core validates and installs it from the adaptive
// runtime; this struct only has to carry the section through to the FFI validator.
type ResponseCacheConfig struct {
	// TTLSeconds is how long a stored answer stays reusable, in seconds (> 0).
	// Nil delegates to Rust's default (3600); a pointer to 0 is rejected.
	TTLSeconds *uint64 `json:"ttl_seconds,omitempty"`
	// Namespace is a required, non-empty cache trust domain folded into every
	// key. One configured cache must not span mutually untrusted tenants or
	// upstreams; the empty constructor value is rejected at validation.
	Namespace string `json:"namespace,omitempty"`
	// Priority is the execution-intercept priority; lower runs first/outermost.
	// Nil delegates to Rust's default (50); a pointer to 0 selects outermost.
	Priority *int32 `json:"priority,omitempty"`
	// BypassRate is the probability in [0.0, 1.0] of skipping the cache and running live.
	BypassRate float64 `json:"bypass_rate,omitempty"`
	// CacheNondeterministic lets requests that are not explicitly deterministic
	// use the cache (default false).
	CacheNondeterministic bool `json:"cache_nondeterministic"`
	// KeyStrategy is the key strategy. Only "exact_request" is supported.
	KeyStrategy string `json:"key_strategy,omitempty"`
	// HeaderAllowlist lists request headers folded into the key; never auth headers.
	HeaderAllowlist []string `json:"header_allowlist,omitempty"`
	// Backend selects the cache's own storage backend (distinct from the adaptive
	// state backend). Defaults to in-memory when nil.
	Backend *ResponseCacheBackendConfig `json:"backend,omitempty"`
	// Tools configures the optional tool-result cache.
	Tools *ResponseCacheToolsConfig `json:"tools,omitempty"`
}

// ResponseCacheToolsConfig configures caching for read-only, stable tools.
type ResponseCacheToolsConfig struct {
	Enabled   bool                                 `json:"enabled,omitempty"`
	Priority  int32                                `json:"priority"`
	Default   *ResponseCacheToolClass              `json:"default,omitempty"`
	Classes   map[string]ResponseCacheToolClass    `json:"classes,omitempty"`
	Overrides map[string]ResponseCacheToolOverride `json:"overrides,omitempty"`
}

// ResponseCacheToolClass defines a shared tool-cache policy.
type ResponseCacheToolClass struct {
	Cacheable  bool     `json:"cacheable,omitempty"`
	TTLSeconds *uint64  `json:"ttl_seconds,omitempty"`
	BypassRate *float64 `json:"bypass_rate,omitempty"`
	ArgSkip    []string `json:"arg_skip,omitempty"`
	Members    []string `json:"members,omitempty"`
}

// ResponseCacheToolOverride refines a resolved tool-cache policy.
type ResponseCacheToolOverride struct {
	Cacheable   *bool     `json:"cacheable,omitempty"`
	TTLSeconds  *uint64   `json:"ttl_seconds,omitempty"`
	BypassRate  *float64  `json:"bypass_rate,omitempty"`
	ToolVersion *string   `json:"tool_version,omitempty"`
	ArgSkip     *[]string `json:"arg_skip,omitempty"`
}

// ResponseCacheBackendConfig selects the response-cache backend kind and options.
type ResponseCacheBackendConfig struct {
	// Kind is "in_memory" or "redis" (redis needs the redis-backend build feature).
	Kind string `json:"kind"`
	// Config holds backend-specific options (in_memory: max_bytes;
	// redis: url/key_prefix).
	Config map[string]any `json:"config,omitempty"`
}

// AdaptiveComponentSpec wraps one adaptive config as a top-level plugin component.
type AdaptiveComponentSpec struct {
	Enabled bool           `json:"enabled,omitempty"`
	Config  AdaptiveConfig `json:"config"`
}

// NewAdaptiveConfig returns a default adaptive config with version 1.
func NewAdaptiveConfig() AdaptiveConfig {
	return AdaptiveConfig{Version: 1}
}

// NewInMemoryAdaptiveBackend returns an in-memory adaptive backend spec.
func NewInMemoryAdaptiveBackend() AdaptiveBackendSpec {
	return AdaptiveBackendSpec{
		Kind:   "in_memory",
		Config: map[string]any{},
	}
}

// NewRedisAdaptiveBackend returns a Redis adaptive backend spec.
func NewRedisAdaptiveBackend(url, keyPrefix string) AdaptiveBackendSpec {
	return AdaptiveBackendSpec{
		Kind: "redis",
		Config: map[string]any{
			"url":        url,
			"key_prefix": keyPrefix,
		},
	}
}

// NewTelemetryConfig returns default built-in adaptive telemetry settings.
func NewTelemetryConfig() TelemetryConfig {
	return TelemetryConfig{}
}

// NewAdaptiveHintsConfig returns default adaptive hints injection settings.
func NewAdaptiveHintsConfig() AdaptiveHintsConfig {
	return AdaptiveHintsConfig{
		Priority:       100,
		InjectHeader:   true,
		InjectBodyPath: "nvext.agent_hints",
	}
}

// NewToolParallelismConfig returns default adaptive tool scheduling settings.
func NewToolParallelismConfig() ToolParallelismConfig {
	return ToolParallelismConfig{
		Priority: 100,
		Mode:     "observe_only",
	}
}

// NewAcgStabilityThresholds returns default ACG stability thresholds.
func NewAcgStabilityThresholds() AcgStabilityThresholds {
	return AcgStabilityThresholds{
		StableThreshold:                  0.95,
		SemiStableThreshold:              0.50,
		MinObservationsForFullConfidence: 20,
	}
}

// NewAcgConfig returns default adaptive cache governor settings.
func NewAcgConfig() AcgConfig {
	thresholds := NewAcgStabilityThresholds()
	return AcgConfig{
		Provider:            "passthrough",
		ObservationWindow:   100,
		Priority:            50,
		StabilityThresholds: &thresholds,
	}
}

// NewResponseCacheConfig returns default response cache settings, mirroring
// the Rust ResponseCacheConfig defaults (exact-request keying with nondeterministic
// caching disabled). Set Namespace to one cache trust domain before validation.
// Backend is left nil so the core applies its in-memory default; set it for redis
// or to tune the in-memory budget.
func NewResponseCacheConfig() ResponseCacheConfig {
	ttlSeconds := uint64(3600)
	priority := int32(50)
	return ResponseCacheConfig{
		TTLSeconds:            &ttlSeconds,
		Priority:              &priority,
		CacheNondeterministic: false,
		KeyStrategy:           "exact_request",
	}
}

// NewInMemoryResponseCacheBackend returns an in-memory response-cache backend spec.
func NewInMemoryResponseCacheBackend() ResponseCacheBackendConfig {
	return ResponseCacheBackendConfig{
		Kind:   "in_memory",
		Config: map[string]any{},
	}
}

// NewRedisResponseCacheBackend returns a Redis response-cache backend spec.
func NewRedisResponseCacheBackend(url, keyPrefix string) ResponseCacheBackendConfig {
	return ResponseCacheBackendConfig{
		Kind: "redis",
		Config: map[string]any{
			"url":        url,
			"key_prefix": keyPrefix,
		},
	}
}

// NewResponseCacheToolsConfig returns a disabled tool-result cache config.
func NewResponseCacheToolsConfig() ResponseCacheToolsConfig {
	return ResponseCacheToolsConfig{
		Priority: 50,
	}
}

// NewAdaptiveComponentSpec wraps adaptive config as an enabled top-level component.
func NewAdaptiveComponentSpec(config AdaptiveConfig) AdaptiveComponentSpec {
	return AdaptiveComponentSpec{
		Enabled: true,
		Config:  config,
	}
}

// PluginComponent converts the adaptive component wrapper into the shared plugin shape.
func (spec AdaptiveComponentSpec) PluginComponent() PluginComponentSpec {
	return PluginComponentSpec{
		Kind:    AdaptivePluginKind,
		Enabled: spec.Enabled,
		Config:  mustConfigMap(spec.Config),
	}
}

// AdaptiveComponent converts adaptive config directly into a shared plugin component.
//
// Prefer NewAdaptiveComponentSpec(config).PluginComponent() when the adaptive
// component wrapper itself is part of the public surface you want to expose.
func AdaptiveComponent(config AdaptiveConfig) PluginComponentSpec {
	return NewAdaptiveComponentSpec(config).PluginComponent()
}

func mustConfigMap(value any) map[string]any {
	payload, err := json.Marshal(value)
	if err != nil {
		panic(err)
	}
	var out map[string]any
	if err := json.Unmarshal(payload, &out); err != nil {
		panic(err)
	}
	if out == nil {
		return map[string]any{}
	}
	return out
}

File: python/nemo_relay/adaptive.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Adaptive plugin configuration helpers.

Adaptive is configured as a single flat top-level plugin component. Hosted
plugins remain separate top-level components managed through ``nemo_relay.plugin``.
"""

from __future__ import annotations

from dataclasses import dataclass, field, fields, is_dataclass
from typing import Literal, Protocol, TypedDict, cast

from nemo_relay import Json, JsonObject, UnsupportedBehavior
from nemo_relay._native import AdaptiveRuntime as AdaptiveRuntime
from nemo_relay._native import build_cache_telemetry_event as _build_cache_telemetry_event
from nemo_relay._native import set_latency_sensitivity as _set_latency_sensitivity
from nemo_relay._native import validate_adaptive_config as _validate_adaptive_config


class _ConfigDiagnosticRequired(TypedDict):
    level: Literal["warning", "error"]
    code: str
    message: str


class ConfigDiagnostic(_ConfigDiagnosticRequired, total=False):
    """One adaptive validation diagnostic."""

    component: str
    field: str


class ConfigReport(TypedDict):
    """Validation report for adaptive configuration."""

    diagnostics: list[ConfigDiagnostic]


class _SupportsToDict(Protocol):
    def to_dict(self) -> JsonObject: ...


def _normalize(value: object) -> Json:
    if hasattr(value, "to_dict"):
        return cast(_SupportsToDict, value).to_dict()
    if is_dataclass(value) and not isinstance(value, type):
        return {
            field_info.name: _normalize(field_value)
            for field_info in fields(value)
            if (field_value := getattr(value, field_info.name)) is not None
        }
    if isinstance(value, list):
        return [_normalize(item) for item in value]
    if isinstance(value, dict):
        return {cast(str, key): _normalize(val) for key, val in value.items() if val is not None}
    return cast(Json, value)


def _normalize_object(value: object) -> JsonObject:
    return cast(JsonObject, _normalize(value))


@dataclass(slots=True)
class ConfigPolicy:
    """Policy for unsupported adaptive configuration.

    Args:
        unknown_component: How to handle unknown component kinds.
        unknown_field: How to handle unknown adaptive config fields.
        unsupported_value: How to handle known fields with unsupported values.
    """

    unknown_component: UnsupportedBehavior = "warn"
    unknown_field: UnsupportedBehavior = "warn"
    unsupported_value: UnsupportedBehavior = "error"

    def to_dict(self) -> JsonObject:
        """Serialize this policy to the canonical JSON object shape."""
        return {
            "unknown_component": self.unknown_component,
            "unknown_field": self.unknown_field,
            "unsupported_value": self.unsupported_value,
        }


@dataclass(slots=True)
class BackendSpec:
    """Adaptive state backend selection.

    Args:
        kind: Backend kind string such as ``"in_memory"`` or ``"redis"``.
        config: Backend-specific JSON object.
    """

    kind: str
    config: JsonObject = field(default_factory=dict)

    @staticmethod
    def in_memory() -> "BackendSpec":
        """Return an in-memory adaptive backend spec."""
        return BackendSpec(kind="in_memory")

    @staticmethod
    def redis(url: str, key_prefix: str = "nemo_relay:") -> "BackendSpec":
        """Return a Redis adaptive backend spec."""
        return BackendSpec(kind="redis", config={"url": url, "key_prefix": key_prefix})

    def to_dict(self) -> JsonObject:
        """Serialize this backend spec to the canonical JSON object shape."""
        return {"kind": self.kind, "config": _normalize_object(self.config)}


@dataclass(slots=True)
class StateConfig:
    """Adaptive state configuration.

    Args:
        backend: State backend selection for adaptive features that persist or
            learn over time.
    """

    backend: BackendSpec

    def to_dict(self) -> JsonObject:
        """Serialize this state config to the canonical JSON object shape."""
        return {"backend": _normalize_object(self.backend)}


@dataclass(slots=True)
class TelemetryConfig:
    """Built-in adaptive telemetry subscriber settings.

    Args:
        subscriber_name: Optional subscriber registration name override.
        learners: Enabled learner identifiers.
    """

    subscriber_name: str | None = None
    learners: list[str] = field(default_factory=list)

    def to_dict(self) -> JsonObject:
        """Serialize this telemetry config to the canonical JSON object shape."""
        return _normalize_object(
            {
                "subscriber_name": self.subscriber_name,
                "learners": self.learners,
            }
        )


@dataclass(slots=True)
class AdaptiveHintsConfig:
    """Built-in adaptive hints injection settings.

    Args:
        priority: Intercept priority. Lower values run first.
        break_chain: Whether to stop later request intercepts after this one.
        inject_header: Whether to inject the adaptive hints HTTP header.
        inject_body_path: JSON body path used when injecting request-body hints.
    """

    priority: int = 100
    break_chain: bool = False
    inject_header: bool = True
    inject_body_path: str = "nvext.agent_hints"

    def to_dict(self) -> JsonObject:
        """Serialize this adaptive-hints config to the canonical JSON object shape."""
        return _normalize_object(
            {
                "priority": self.priority,
                "break_chain": self.break_chain,
                "inject_header": self.inject_header,
                "inject_body_path": self.inject_body_path,
            }
        )


@dataclass(slots=True)
class ToolParallelismConfig:
    """Built-in adaptive tool scheduling settings.

    Args:
        priority: Intercept priority. Lower values run first.
        mode: Scheduling mode. ``"observe_only"`` records signals without
            changing behavior, while other modes enable stronger adaptive
            scheduling behavior.
    """

    priority: int = 100
    mode: Literal["observe_only", "inject_hints", "schedule"] = "observe_only"

    def to_dict(self) -> JsonObject:
        """Serialize this tool-parallelism config to the canonical JSON object shape."""
        return _normalize_object({"priority": self.priority, "mode": self.mode})


@dataclass(slots=True)
class AcgStabilityThresholds:
    """Prompt-stability classification thresholds for ACG.

    Args:
        stable_threshold: Minimum effective score classified as stable.
        semi_stable_threshold: Minimum effective score classified as semi-stable.
        min_observations_for_full_confidence: Observation count required to
            reach full confidence.
    """

    stable_threshold: float = 0.95
    semi_stable_threshold: float = 0.50
    min_observations_for_full_confidence: int = 20

    def to_dict(self) -> JsonObject:
        """Serialize these ACG stability thresholds to the canonical JSON object shape."""
        return _normalize_object(
            {
                "stable_threshold": self.stable_threshold,
                "semi_stable_threshold": self.semi_stable_threshold,
                "min_observations_for_full_confidence": self.min_observations_for_full_confidence,
            }
        )


@dataclass(slots=True)
class AcgConfig:
    """Adaptive Cache Governor settings.

    Args:
        provider: Provider cache plugin name.
        observation_window: Rolling PromptIR observation window size.
        priority: LLM execution intercept priority.
        stability_thresholds: Prompt-stability classification thresholds.
    """

    provider: Literal["anthropic", "openai", "passthrough"] = "passthrough"
    observation_window: int = 100
    priority: int = 50
    stability_thresholds: AcgStabilityThresholds | None = field(default_factory=AcgStabilityThresholds)

    def to_dict(self) -> JsonObject:
        """Serialize this ACG config to the canonical JSON object shape."""
        return _normalize_object(
            {
                "provider": self.provider,
                "observation_window": self.observation_window,
                "priority": self.priority,
                "stability_thresholds": _normalize(self.stability_thresholds),
            }
        )


@dataclass(slots=True)
class ToolClass:
    """One tool caching class (also the shape of the ``default`` default bucket).

    Args:
        cacheable: Whether tools in this class may be served from cache. Off by
            default — a hit suppresses the real call, so caching must be opted in.
        ttl_seconds: TTL for this class; inherits ``response_cache.ttl_seconds``
            when ``None``.
        bypass_rate: Live-rerun probability for this class; inherits
            ``response_cache.bypass_rate`` when ``None``.
        arg_skip: Argument keys dropped before keying (default empty: key on all args).
        members: Tool names in this class (unused for the ``default`` bucket).
            Names may use ``*`` wildcards; an exact member wins over any
            wildcard match, the most-specific pattern wins among wildcards, and
            unmatched tools fall to ``default``.
    """

    cacheable: bool = False
    ttl_seconds: int | None = None
    bypass_rate: float | None = None
    arg_skip: list[str] = field(default_factory=list)
    members: list[str] = field(default_factory=list)

    def to_dict(self) -> JsonObject:
        """Serialize this tool class to the canonical JSON object shape."""
        return _normalize_object(
            {
                "cacheable": self.cacheable,
                "ttl_seconds": self.ttl_seconds,
                "bypass_rate": self.bypass_rate,
                "arg_skip": self.arg_skip,
                "members": self.members,
            }
        )


@dataclass(slots=True)
class ToolOverride:
    """Per-tool refinement applied on top of the tool's resolved class.

    Args:
        cacheable: Overrides the class ``cacheable`` for just this tool.
        ttl_seconds: Overrides the class TTL for just this tool.
        bypass_rate: Overrides the class bypass rate for just this tool.
        tool_version: Version string folded into the key so a deployment can bust
            stale entries before their TTL.
        arg_skip: Replaces the class ``arg_skip`` when not ``None`` (``None``
            inherits the class list; ``[]`` clears it).
    """

    cacheable: bool | None = None
    ttl_seconds: int | None = None
    bypass_rate: float | None = None
    tool_version: str | None = None
    arg_skip: list[str] | None = None

    def to_dict(self) -> JsonObject:
        """Serialize this tool override to the canonical JSON object shape."""
        return _normalize_object(
            {
                "cacheable": self.cacheable,
                "ttl_seconds": self.ttl_seconds,
                "bypass_rate": self.bypass_rate,
                "tool_version": self.tool_version,
                "arg_skip": self.arg_skip,
            }
        )


@dataclass(slots=True)
class ToolCacheConfig:
    """Opt-in tool-result cache settings.

    A separate surface under ``response_cache`` keyed on tool name + arguments and
    gated by user-declared safety classes. Off until ``enabled`` is set; any tool
    not listed in a class falls into ``default``, which defaults to not cached.

    Args:
        enabled: Master switch for the tool surface. Off by default.
        priority: Tool execution-intercept priority. Lower runs first/outermost.
        default: Policy for tools not listed in any class (defaults to not cached).
        classes: Named tool classes, each with its own policy and member list.
        overrides: Per-tool refinements applied on top of the resolved class.
            Keys may be exact tool names or ``*`` patterns; an exact key wins
            outright, then the most-specific matching pattern applies.
    """

    enabled: bool = False
    priority: int = 50
    default: ToolClass = field(default_factory=ToolClass)
    classes: dict[str, ToolClass] = field(default_factory=dict)
    overrides: dict[str, ToolOverride] = field(default_factory=dict)

    def to_dict(self) -> JsonObject:
        """Serialize this tool-cache config to the canonical JSON object shape."""
        return _normalize_object(
            {
                "enabled": self.enabled,
                "priority": self.priority,
                "default": _normalize(self.default),
                "classes": {name: _normalize(cls) for name, cls in self.classes.items()},
                "overrides": {name: _normalize(ov) for name, ov in self.overrides.items()},
            }
        )


@dataclass(slots=True)
class ResponseCacheConfig:
    """Opt-in LLM response cache (exact-match) settings.

    This is a section of the adaptive component, not a standalone plugin kind.
    When present, the adaptive plugin installs the response-cache execution
    intercept that reuses an earlier answer for a repeated managed LLM call.

    Args:
        ttl_seconds: How long a stored answer stays reusable, in seconds.
        namespace: Required non-empty cache trust domain folded into every key.
            One configured cache must not span mutually untrusted tenants or upstreams;
            the empty default is an unconfigured sentinel rejected at validation.
        priority: Execution-intercept priority. Lower runs first/outermost.
        bypass_rate: Probability in ``[0.0, 1.0]`` of skipping the cache and running live.
        cache_nondeterministic: Cache nondeterministic requests too; ``False``
            caches only requests explicitly pinned deterministic (``temperature`` = 0).
        key_strategy: Key strategy. Only ``"exact_request"`` is supported.
        header_allowlist: Request headers folded into the key; never auth headers.
        backend: Cache storage backend (``in_memory`` or ``redis``).
        tools: Opt-in tool-result cache; ``None`` leaves it off.
    """

    ttl_seconds: int = 3600
    namespace: str = ""
    priority: int = 50
    bypass_rate: float = 0.0
    cache_nondeterministic: bool = False
    key_strategy: str = "exact_request"
    header_allowlist: list[str] = field(default_factory=list)
    backend: BackendSpec = field(default_factory=BackendSpec.in_memory)
    tools: ToolCacheConfig | None = None

    def to_dict(self) -> JsonObject:
        """Serialize this response-cache config to the canonical JSON object shape."""
        return _normalize_object(
            {
                "ttl_seconds": self.ttl_seconds,
                "namespace": self.namespace,
                "priority": self.priority,
                "bypass_rate": self.bypass_rate,
                "cache_nondeterministic": self.cache_nondeterministic,
                "key_strategy": self.key_strategy,
                "header_allowlist": self.header_allowlist,
                "backend": _normalize(self.backend),
                "tools": _normalize(self.tools),
            }
        )


@dataclass(slots=True)
class AdaptiveConfig:
    """Canonical config document for the top-level adaptive component.

    Args:
        version: Adaptive config schema version.
        agent_id: Optional explicit agent identifier for learned state.
        state: Adaptive state backend configuration.
        telemetry: Built-in adaptive telemetry settings.
        adaptive_hints: Built-in LLM hint-injection settings.
        tool_parallelism: Built-in tool scheduling settings.
        acg: Adaptive Cache Governor settings.
        policy: Unsupported-config policy applied within the adaptive config.
        response_cache: Opt-in LLM response cache settings.

    Behavior:
        This document configures only the adaptive component. Plugins are
        configured separately through top-level plugin components.
    """

    version: int = 1
    agent_id: str | None = None
    state: StateConfig | None = None
    telemetry: TelemetryConfig | None = None
    adaptive_hints: AdaptiveHintsConfig | None = None
    tool_parallelism: ToolParallelismConfig | None = None
    acg: AcgConfig | None = None
    policy: ConfigPolicy = field(default_factory=ConfigPolicy)
    response_cache: ResponseCacheConfig | None = None

    def to_dict(self) -> JsonObject:
        """Serialize this adaptive config to the canonical JSON object shape."""
        return {
            "version": self.version,
            "agent_id": self.agent_id,
            "state": _normalize(self.state),
            "telemetry": _normalize(self.telemetry),
            "adaptive_hints": _normalize(self.adaptive_hints),
            "tool_parallelism": _normalize(self.tool_parallelism),
            "acg": _normalize(self.acg),
            "response_cache": _normalize(self.response_cache),
            "policy": self.policy.to_dict(),
        }


ADAPTIVE_PLUGIN_KIND = "adaptive"


@dataclass(slots=True)
class ComponentSpec:
    """Top-level adaptive component wrapper.

    Args:
        config: ``AdaptiveConfig`` or an equivalent JSON object.
        enabled: Whether the adaptive component should be activated.

    Behavior:
        The component kind is always ``"adaptive"``.
    """

    config: AdaptiveConfig | JsonObject
    enabled: bool = True

    def to_dict(self) -> JsonObject:
        """Serialize this component to the canonical plugin shape."""
        return {
            "kind": ADAPTIVE_PLUGIN_KIND,
            "enabled": self.enabled,
            "config": _normalize_object(self.config),
        }


def validate_config(config: AdaptiveConfig | JsonObject) -> ConfigReport:
    """Validate an adaptive config document without constructing a runtime."""
    return cast(ConfigReport, _validate_adaptive_config(_normalize_object(config)))


def build_cache_telemetry_event(
    *,
    provider: str,
    request_id: str,
    usage: JsonObject | None = None,
    request_facts: JsonObject | None = None,
    agent_id: str,
    template_version: str,
    toolset_hash: str,
    model_family: str,
    tenant_scope: str,
    timestamp: str | None = None,
) -> JsonObject | None:
    """Build one canonical cache telemetry event from usage plus request facts."""
    return cast(
        JsonObject | None,
        _build_cache_telemetry_event(
            provider=provider,
            request_id=request_id,
            usage=usage,
            request_facts=request_facts,
            agent_id=agent_id,
            template_version=template_version,
            toolset_hash=toolset_hash,
            model_family=model_family,
            tenant_scope=tenant_scope,
            timestamp=timestamp,
        ),
    )


def set_latency_sensitivity(level: int) -> None:
    """Set a request-local latency-sensitivity hint.

    Args:
        level: Positive integer sensitivity value for the current execution
            context.

    Returns:
        `None`.

    Behavior:
        This is an execution-time hint for the current request/scope context,
        not persistent adaptive configuration. The native adaptive layer stores
        this as a positive integer.
    """
    _set_latency_sensitivity(level)


__all__ = [
    "AcgConfig",
    "AcgStabilityThresholds",
    "AdaptiveConfig",
    "AdaptiveHintsConfig",
    "ADAPTIVE_PLUGIN_KIND",
    "BackendSpec",
    "ConfigDiagnostic",
    "ConfigPolicy",
    "ConfigReport",
    "ComponentSpec",
    "ResponseCacheConfig",
    "StateConfig",
    "TelemetryConfig",
    "ToolCacheConfig",
    "ToolClass",
    "ToolOverride",
    "ToolParallelismConfig",
    "set_latency_sensitivity",
    "UnsupportedBehavior",
]
